### PR TITLE
Feature/add pg source

### PR DESCRIPTION
**Pull request by xuzh2024**

### CODE REVIEW

File: indexer/jobs/source_job/pg_source_job.py [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+            )
+
+        query_filter = and_(
+            query_filter, Transactions.block_number >= start_block, Transactions.block_number <= end_block
+        )
+
+        session = self._service.get_service_session()
+        try:
+            transactions = (
+                session.query(Transactions).filter(query_filter).order_by(*Transactions.__query_order__).all()
```
Comment (shanshuo0918) [inline indexer/jobs/source_job/pg_source_job.py:224]: Same question here. Need to test performance

Comment (shanshuo0918) [reply indexer/jobs/source_job/pg_source_job.py:224]: Check whether enough index.

File: indexer/jobs/source_job/pg_source_job.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+        self.domain_mapping = defaultdict(dict)
+        self.pg_datas = defaultdict(list)
+        self._filters = flatten(kwargs.get("filters", []))
+        self._is_filter = kwargs.get("is_filter", False)
+        self._specification = AlwaysFalseSpecification() if self._is_filter else AlwaysTrueSpecification()
+
+        if self._service is None:
+            raise FastShutdownError("-pg or --postgres-url is required to run PGSourceJob")
+
+        self.build_dependency = {}
```
Comment (shanshuo0918) [inline indexer/jobs/source_job/pg_source_job.py:59]: What does this build_dependency do?

File: cli/stream.py [hunk 3]
```diff
@@ -343,18 +373,12 @@ def stream(
...
-                raise click.ClickException(f"Output type {output_type} is not supported")
-            parse_output_types.add(domain_dict[output_type])
+        output_types = generate_dataclass_type_list_from_parameter(output_types, "output")
 
-        if not output_types:
-            raise click.ClickException("No output types provided")
-        output_types = list(set(parse_output_types))
+    if source_types is None and source_path.startwith("postgresql://"):
+        source_types = "block,transaction,log"
+    if source_types:
```
Comment (shanshuo0918) [inline cli/stream.py:380]: Should we add this extra step or we should enforce user to list all the source types?

File: cli/stream.py [hunk 4]
```diff
@@ -375,6 +400,9 @@ def stream(
         batch_web3_provider=ThreadLocalProxy(lambda: get_provider_from_uri(provider_uri, batch=False)),
         job_scheduler=job_scheduler,
         sync_recorder=create_recorder(sync_recorder, config),
+        limit_reader=create_limit_reader(
```
Comment (shanshuo0918) [inline cli/stream.py:403]: what if user put other value in source_path?

File: indexer/jobs/source_job/pg_source_job.py [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+                Logs.transaction_hash.in_(
+                    [bytes.fromhex(transaction_hash[2:]) for transaction_hash in set(log_filter["transaction_hash"])]
+                ),
+            )
+
+        query_filter = and_(query_filter, Logs.block_number >= start_block, Logs.block_number <= end_block)
+
+        session = self._service.get_service_session()
+        try:
+            logs = session.query(Logs).filter(query_filter).order_by(*Logs.__query_order__).all()
```
[the remark this answers is not in the document]
Comment (xuzh2024) [reply indexer/jobs/source_job/pg_source_job.py:185]: I will add partition key "timestamp" to the query to enhance its performance.
The order statement would only sort the results set, and I use it for uniform processing.
If the order statement has a significant negative impact on execution, I will do in memory ordering.

File: indexer/jobs/source_job/pg_source_job.py (hunk 1 above)
Comment (xuzh2024) [reply indexer/jobs/source_job/pg_source_job.py:224]: Index has been checked.
I will add partition key "timestamp" to the query to enhance its performance.

File: indexer/jobs/source_job/pg_source_job.py (hunk 2 above)
Comment (xuzh2024) [reply indexer/jobs/source_job/pg_source_job.py:59]: This is legacy code in development and will be removed in the next commit.

File: cli/stream.py (hunk 3 above)
Comment (xuzh2024) [reply cli/stream.py:380]: Instead of this step, I set a default value to “--source-types”

File: cli/stream.py (hunk 4 above)
Comment (xuzh2024) [reply cli/stream.py:403]: In unexpected cases, limit_reader will request the latest block from rpc.
if source_path specify a csv_source, parameter checking would ensure that the rpc_limit_reader does not interfere with the program

**Review by shanshuo0918 — APPROVED** [review]

**Pull request merged**